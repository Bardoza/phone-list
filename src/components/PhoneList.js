import React, { Component } from "react";
import PhoneItem from "./PhoneItem";

class PhoneList extends Component {
  state = {
    types: ["Home", "Mobile", "Work", "Other"],
    phones: [
      {
        number: "4777153366",
        type: "Work"
      }
    ]
  };

  getAvailableTypes = () => {
    return this.state.types.filter(
      type => !this.state.phones.some(phone => phone.type === type)
    );
  };

  render() {
    const availableTypes = this.getAvailableTypes();
    return (
      <div>
        <div className="phone-container">
          {this.state.phones.map((p, ix) => (
            <PhoneItem
              key={ix}
              index={ix}
              {...p}
              types={this.state.types}
              availableTypes={availableTypes}
            />
          ))}
        </div>
        <div>
          <button>Log</button>
          <button>Add Another</button>
        </div>
      </div>
    );
  }
}

export default PhoneList;
