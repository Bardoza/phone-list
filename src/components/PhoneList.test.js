import React from 'react'
import { shallow } from 'enzyme'
import PhoneList, { defaultPhone } from './PhoneList';

/**
 *  Proptypes
 *  title: PropTypes.string,
 */

let phoneList;




beforeEach( () => {
    phoneList = <PhoneList title="List 1" />

})
describe('phoneList', () => {
    
    it("Render correctly", () => {
        expect(shallow(phoneList)).toMatchSnapshot()  
    })

    it("should set defautl state on init", () =>{
        const wrapper = shallow(phoneList)
        expect(wrapper.state().phones).toEqual([defaultPhone])
    })


    it("should return an array with the unused phone types", () =>{
        const wrapper = shallow(phoneList)
        const inst = wrapper.instance()
        wrapper.setState({
            phones: [{
                type:"Work",
                number:"47736541234"
            }]
        })
        expect(inst.getAvailableTypes()).toEqual([
            'Home','Mobile', "Other"
        ])
    })
    
    it("should render the footer  buttons", () => {
        const wrapper = shallow(phoneList)
        expect(wrapper.find(".button")).toHaveLength(2)
    })
    
    it("should add a new phone to the list", () => {
        const wrapper = shallow(phoneList)
        expect(wrapper.find("PhoneItem")).toHaveLength(1)
        wrapper.find('.button-primary').simulate("click")
        expect(wrapper.find("PhoneItem")).toHaveLength(2)
    })
    
    
    
    it("should not add a new phone if there are no free types", () =>{
        const wrapper = shallow(phoneList)
        wrapper.setState({
            phones: [{
                type:"Home",
                number: "48726345397"
            },{
                type:"Work",
                number:"47736541234"
            },
            {
                type:"Mobile",
                number:"4498736162"
            },
            {
                type:"Other",
                number:"4773971635"
            }]
        })

        expect(wrapper.find("PhoneItem")).toHaveLength(4)
        wrapper.find('.button-primary').simulate("click")
        expect(wrapper.find("PhoneItem")).toHaveLength(4)


    })
    
})