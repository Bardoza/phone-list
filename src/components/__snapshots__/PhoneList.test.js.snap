// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`phoneList Render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PhoneList
    title="List 1"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3>
          List 1
</h3>,
        <div
          className="phone-list"
>
          <PhoneItem
                    availableTypes={
                              Array [
                                        "Mobile",
                                        "Work",
                                        "Other",
                                      ]
                    }
                    index={0}
                    number=""
                    onChange={[Function]}
                    onDelete={[Function]}
                    type="Home"
                    types={
                              Array [
                                        "Home",
                                        "Mobile",
                                        "Work",
                                        "Other",
                                      ]
                    }
          />
</div>,
        <div
          className="phone-footer"
>
          <button
                    className="button"
                    onClick={[Function]}
          >
                    Log
          </button>
          <button
                    className="button button-primary"
                    onClick={[Function]}
          >
                    Add Another
          </button>
</div>,
      ],
      "className": "phone-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "List 1",
        },
        "ref": null,
        "rendered": "List 1",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <PhoneItem
              availableTypes={
                            Array [
                                          "Mobile",
                                          "Work",
                                          "Other",
                                        ]
              }
              index={0}
              number=""
              onChange={[Function]}
              onDelete={[Function]}
              type="Home"
              types={
                            Array [
                                          "Home",
                                          "Mobile",
                                          "Work",
                                          "Other",
                                        ]
              }
/>,
          ],
          "className": "phone-list",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "availableTypes": Array [
                "Mobile",
                "Work",
                "Other",
              ],
              "index": 0,
              "number": "",
              "onChange": [Function],
              "onDelete": [Function],
              "type": "Home",
              "types": Array [
                "Home",
                "Mobile",
                "Work",
                "Other",
              ],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="button"
              onClick={[Function]}
>
              Log
</button>,
            <button
              className="button button-primary"
              onClick={[Function]}
>
              Add Another
</button>,
          ],
          "className": "phone-footer",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Log",
              "className": "button",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Log",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Add Another",
              "className": "button button-primary",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Add Another",
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3>
            List 1
</h3>,
          <div
            className="phone-list"
>
            <PhoneItem
                        availableTypes={
                                    Array [
                                                "Mobile",
                                                "Work",
                                                "Other",
                                              ]
                        }
                        index={0}
                        number=""
                        onChange={[Function]}
                        onDelete={[Function]}
                        type="Home"
                        types={
                                    Array [
                                                "Home",
                                                "Mobile",
                                                "Work",
                                                "Other",
                                              ]
                        }
            />
</div>,
          <div
            className="phone-footer"
>
            <button
                        className="button"
                        onClick={[Function]}
            >
                        Log
            </button>
            <button
                        className="button button-primary"
                        onClick={[Function]}
            >
                        Add Another
            </button>
</div>,
        ],
        "className": "phone-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "List 1",
          },
          "ref": null,
          "rendered": "List 1",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <PhoneItem
                availableTypes={
                                Array [
                                                "Mobile",
                                                "Work",
                                                "Other",
                                              ]
                }
                index={0}
                number=""
                onChange={[Function]}
                onDelete={[Function]}
                type="Home"
                types={
                                Array [
                                                "Home",
                                                "Mobile",
                                                "Work",
                                                "Other",
                                              ]
                }
/>,
            ],
            "className": "phone-list",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "availableTypes": Array [
                  "Mobile",
                  "Work",
                  "Other",
                ],
                "index": 0,
                "number": "",
                "onChange": [Function],
                "onDelete": [Function],
                "type": "Home",
                "types": Array [
                  "Home",
                  "Mobile",
                  "Work",
                  "Other",
                ],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                className="button"
                onClick={[Function]}
>
                Log
</button>,
              <button
                className="button button-primary"
                onClick={[Function]}
>
                Add Another
</button>,
            ],
            "className": "phone-footer",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Log",
                "className": "button",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Log",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Add Another",
                "className": "button button-primary",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Add Another",
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
